File: java/src/hackerrank/CountingValleys.java
package hackerrank;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

// https://www.hackerrank.com/challenges/counting-valleys/problem
public class CountingValleys {

	public static int countingValleys(int n, String s) {
		int countOfValleys = 0, path = 0, seaLevel = 0;
		boolean enteredValley = false;
		for (int i = 0; i < n; i++) {
			Character letter = s.charAt(i);
			if (letter.equals('U')) path += 1;
			else if (letter.equals('D')) path -= 1;
			if (path == seaLevel && enteredValley)
				countOfValleys++;
			enteredValley = path < 0;
		}
		return countOfValleys;
	}


	@Test
	@DisplayName("Test given sample")
	public void testCountingValleys() {
		int n = 8; String s = "UDDDUDUU";
		int result = countingValleys(n, s);
		assertEquals(1, result);		
	}
	
}

File: javascript/hackerrank/counting-valleys.js
// https://www.hackerrank.com/challenges/counting-valleys/problem

const countingValleys = (n, s) => {
    let countOfValleys = 0, path = 0, enteredValley = false;
    const seaLevel = 0;		
    for (let i = 0; i < n; i++) {
        const letter = s.charAt(i);
        if (letter === 'U') path += 1;
        else if (letter === 'D') path -= 1;
        if (path === seaLevel && enteredValley)
            countOfValleys++;
        enteredValley = path < 0;
    }
    return countOfValleys;
}


test("countingValleys", () => {
	const n = 8, s = "UDDDUDUU";
	const result = countingValleys(n, s);
	expect(result).toBe(1);		
})